syntax = "proto3";

package wallet;

service WalletService {
    rpc ReadBalance(ReadBalanceRequest) returns (ReadBalanceResponse);
    rpc CreatePaymentOrder(CreatePaymentOrderRequest) returns (CreatePaymentOrderResponse);
    rpc Transfer(TransferRequest) returns (TransferResponse);
    rpc Shutdown(ShutdownRequest) returns (ShutdownResponse);
}

message ReadBalanceRequest {
    string wallet_id = 1;
}

message ReadBalanceResponse {
    int32 balance = 1;
}

message CreatePaymentOrderRequest {
    string wallet_id = 1;
    int32 amount = 2;
}

message CreatePaymentOrderResponse {
    int32 order_id = 1;
    int32 status = 2;
}

message TransferRequest {
    int32 order_id = 1;
    int32 amount = 2;
    string wallet_id = 3;
}

message TransferResponse {
    int32 status = 1;
}

message ShutdownRequest {}

message ShutdownResponse {
    int32 remaining_orders = 1;
    repeated Wallet wallets = 2;
}

message Wallet {
    string wallet_id = 1;
    int32 balance = 2;
}
